### ISSUE TITLE
significant_text aggregation reversed include/exclude behaviour

### ISSUE THREAD

**Issue by danelowe** [elastic/elasticsearch#62266]
**Elasticsearch version**  7.9.1

**Plugins installed**: [repository-gcs]

**JVM version** (`java -version`):

java version "14.0.2" 2020-07-14
Java(TM) SE Runtime Environment (build 14.0.2+12-46)
Java HotSpot(TM) 64-Bit Server VM (build 14.0.2+12-46, mixed mode, sharing)

**OS version** (`uname -a` if on a Unix-like system):
Darwin Danes-MBP 19.6.0 Darwin Kernel Version 19.6.0: Thu Jun 18 20:49:00 PDT 2020; root:xnu-6153.141.1~1/RELEASE_X86_64 x86_64

**Description of the problem including expected versus actual behavior**:
When using `significant_text` aggregation, the `include` and `exclude` options seem to have reversed behaviour since 6.3.2

```
GET catalog.review/_search
{"size":0,"from":0,"query":{"bool":{"must":[{"terms":{"entityPkValue":["2764"]}},{"term":{"statusId":1}}]}},"aggregations":{"word_cloud":{"sampler":{"shard_size":100},"aggregations":{"keywords":{"significant_text":{"size":30,"field":"detail.phrases","exclude":"storage"}}}}}}

{
  "took" : 20,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 211,
      "relation" : "eq"
    },
    "max_score" : null,
    "hits" : [ ]
  },
  "aggregations" : {
    "word_cloud" : {
      "doc_count" : 100,
      "keywords" : {
        "doc_count" : 100,
        "bg_count" : 14683,
        "buckets" : [
          {
            "key" : "storage",
            "doc_count" : 30,
            "score" : 2.623606194690266,
            "bg_count" : 452
          }
        ]
      }
    }
  }
}
```

```
GET catalog.review/_search
{"size":0,"from":0,"query":{"bool":{"must":[{"terms":{"entityPkValue":["2764"]}},{"term":{"statusId":1}}]}},"aggregations":{"word_cloud":{"sampler":{"shard_size":100},"aggregations":{"keywords":{"significant_text":{"size":30,"field":"detail.phrases","include":"storage"}}}}}}


{
  "took" : 45,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 211,
      "relation" : "eq"
    },
    "max_score" : null,
    "hits" : [ ]
  },
  "aggregations" : {
    "word_cloud" : {
      "doc_count" : 100,
      "keywords" : {
        "doc_count" : 100,
        "bg_count" : 14683,
        "buckets" : [
          {
            "key" : "cushion",
            "doc_count" : 21,
            "score" : 2.99554603960396,
            "bg_count" : 202
          },
          {
            "key" : "storage box",
            "doc_count" : 19,
            "score" : 2.8046683615819212,
            "bg_count" : 177
          },
          {
            "key" : "box",
            "doc_count" : 32,
            "score" : 2.560343295019157,
            "bg_count" : 522
          },
          {
            "key" : "outdoor",
            "doc_count" : 24,
            "score" : 2.009310638297872,
            "bg_count" : 376
          }
        ]
      }
    }
  }
}
```

I can look into creating a reproducer when I have time if necessary, but it's Friday night here and I figured this one is probably quite easy to diagnose.

**Comment by imotov** [comment]
@danelowe would you mind sharing your analyzer for this field? I spent some time trying to reproduce the issue but everything seems to work as expected.

**Comment by danelowe** [comment]
Config for the index in question.

```
{
  "catalog.review.1600034149.9be7bfa3" : {
    "aliases" : {
      "catalog.review" : { },
      "catalog.review.write" : { }
    },
    "mappings" : {
      "properties" : {
        "createdAt" : {
          "type" : "date"
        },
        "detail" : {
          "type" : "text",
          "fields" : {
            "phrases" : {
              "type" : "text",
              "analyzer" : "phrase_cloud_analyzer",
              "similarity" : "BM25"
            }
          },
          "similarity" : "BM25"
        },
        "entityId" : {
          "type" : "integer"
        },
        "entityPkValue" : {
          "type" : "integer"
        },
        "id" : {
          "type" : "keyword"
        },
        "imageData" : {
          "properties" : {
            "o" : {
              "type" : "text"
            },
            "u" : {
              "type" : "text"
            }
          }
        },
        "location" : {
          "type" : "text",
          "index" : false
        },
        "magentoId" : {
          "type" : "integer"
        },
        "nickname" : {
          "type" : "text",
          "index" : false
        },
        "rating" : {
          "type" : "integer"
        },
        "recommend" : {
          "type" : "boolean"
        },
        "replyComment" : {
          "type" : "text",
          "index" : false
        },
        "statusId" : {
          "type" : "integer"
        },
        "storeId" : {
          "type" : "integer"
        },
        "title" : {
          "type" : "text"
        }
      }
    },
    "settings" : {
      "index" : {
        "number_of_shards" : "1",
        "provided_name" : "catalog.review.1600034149.9be7bfa3",
        "creation_date" : "1600034149606",
        "analysis" : {
          "filter" : {
            "minimal_plural_stemmer" : {
              "name" : "minimal_english",
              "type" : "stemmer"
            },
            "phrase_shingles_token_filter" : {
              "max_shingle_size" : "3",
              "output_unigrams" : "true",
              "type" : "shingle"
            },
            "custom_stop_words" : {
              "type" : "stop",
              "stopwords" : [
                "our",
                "i",
                "have",
                "my",
                "a",
                "is",
                "this",
                "have",
                "has"
              ]
            },
            "remove_shingles_with_stopwords" : {
              "type" : "predicate_token_filter",
              "script" : {
                "source" : "!token.getTerm().toString().contains('_')"
              }
            }
          },
          "analyzer" : {
            "phrase_cloud_analyzer" : {
              "filter" : [
                "lowercase",
                "stop",
                "custom_stop_words",
                "minimal_plural_stemmer",
                "phrase_shingles_token_filter",
                "remove_shingles_with_stopwords"
              ],
              "type" : "custom",
              "tokenizer" : "standard"
            }
          }
        },
        "number_of_replicas" : "1",
        "uuid" : "5Q-yl6_PTS-I5z6ToPdkoA",
        "version" : {
          "created" : "7090199"
        }
      }
    }
  }
}
```

**Comment by imotov** [comment]
I still cannot reproduce it. Here is what I have got so far:

```
DELETE test

PUT test
{
  "mappings": {
    "properties": {
      "val": {
        "type": "keyword"
      },
      "term": {
        "type": "text",
        "analyzer": "phrase_cloud_analyzer",
        "similarity": "BM25"
      }
    }
  },
  "settings": {
    "index": {
      "analysis": {
        "filter": {
          "minimal_plural_stemmer": {
            "name": "minimal_english",
            "type": "stemmer"
          },
          "phrase_shingles_token_filter": {
            "max_shingle_size": "3",
            "output_unigrams": "true",
            "type": "shingle"
          },
          "custom_stop_words": {
            "type": "stop",
            "stopwords": [
              "our",
              "i",
              "have",
              "my",
              "a",
              "is",
              "this",
              "have",
              "has"
            ]
          },
          "remove_shingles_with_stopwords": {
            "type": "predicate_token_filter",
            "script": {
              "source": "!token.getTerm().toString().contains('_')"
            }
          }
        },
        "analyzer": {
          "phrase_cloud_analyzer": {
            "filter": [
              "lowercase",
              "stop",
              "custom_stop_words",
              "minimal_plural_stemmer",
              "phrase_shingles_token_filter",
              "remove_shingles_with_stopwords"
            ],
            "type": "custom",
            "tokenizer": "standard"
          }
        }
      }
    }
  }
}

POST test/_analyze
{
  "field": "term",
  "text": ["cloud storage"]
}


POST test/_bulk
{"index":{}}
{"val": "a", "term": "cloud storage is cool"}
{"index":{}}
{"val": "a", "term": "nice storage box in the cloud"}
{"index":{}}
{"val": "a", "term": "storage all the way down and cloud"}
{"index":{}}
{"val": "a", "term": "storage storage storage"}
{"index":{}}
{"val": "b", "term": "cloud cloud cloud"}
{"index":{}}
{"val": "b", "term": "cool cool"}
{"index":{}}
{"val": "b", "term": "blue box"}
{"index":{}}
{"val": "b", "term": "red box"}


GET test/_search
{
  "size": 0,
  "query": {
    "term": {
      "val": {
        "value": "a"
      }
    }
  }, 
  "aggs": {
    "my_sig_text_excluded": {
      "significant_text": {
        "field": "term",
        "exclude": "storage"
      }
    },
    "my_sig_text_included": {
      "significant_text": {
        "field": "term",
        "include": "storage"
      }
    },
    "my_sig_text": {
      "significant_text": {
        "field": "term"
      }
    }
  }
}
```

It seems to work as I would expect it to work.

**Comment by imotov** [comment]
@danelowe any updates on this one? I think the second example could have been caused by a combination of shingles and sampler agg, not really sure why `storage` is not excluded in the first example, but I still cannot reproduce it.

**Comment by danelowe** [comment]
```
DELETE test

PUT test
{
  "mappings": {
    "properties": {
      "val": {
        "type": "keyword"
      },
       "term" : {
          "type" : "text",
          "fields" : {
            "phrases" : {
              "type" : "text",
              "analyzer" : "phrase_cloud_analyzer",
              "similarity" : "BM25"
            }
          },
          "similarity" : "BM25"
        }
    }
  },
  "settings": {
    "index": {
      "analysis": {
        "filter": {
          "minimal_plural_stemmer": {
            "name": "minimal_english",
            "type": "stemmer"
          },
          "phrase_shingles_token_filter": {
            "max_shingle_size": "3",
            "output_unigrams": "true",
            "type": "shingle"
          },
          "custom_stop_words": {
            "type": "stop",
            "stopwords": [
              "our",
              "i",
              "have",
              "my",
              "a",
              "is",
              "this",
              "have",
              "has"
            ]
          },
          "remove_shingles_with_stopwords": {
            "type": "predicate_token_filter",
            "script": {
              "source": "!token.getTerm().toString().contains('_')"
            }
          }
        },
        "analyzer": {
          "phrase_cloud_analyzer": {
            "filter": [
              "lowercase",
              "stop",
              "custom_stop_words",
              "minimal_plural_stemmer",
              "phrase_shingles_token_filter",
              "remove_shingles_with_stopwords"
            ],
            "type": "custom",
            "tokenizer": "standard"
          }
        }
      }
    }
  }
}

POST test/_analyze
{
  "field": "term",
  "text": ["cloud storage"]
}


POST test/_bulk
{"index":{}}
{"val": "a", "term": "cloud storage is cool"}
{"index":{}}
{"val": "a", "term": "nice storage box in the cloud"}
{"index":{}}
{"val": "a", "term": "storage all the way down and cloud"}
{"index":{}}
{"val": "a", "term": "storage storage storage"}
{"index":{}}
{"val": "b", "term": "cloud cloud cloud"}
{"index":{}}
{"val": "b", "term": "cool cool"}
{"index":{}}
{"val": "b", "term": "blue box"}
{"index":{}}
{"val": "b", "term": "red box"}


GET test/_search
{
  "size": 0,
  "query": {
    "term": {
      "val": {
        "value": "a"
      }
    }
  }, 
  "aggs": {
    "my_sig_text_excluded": {
      "significant_text": {
        "field": "term",
        "exclude": "storage"
      }
    },
    "my_sig_text_included": {
      "significant_text": {
        "field": "term",
        "include": "storage"
      }
    },
    "my_sig_text": {
      "significant_text": {
        "field": "term"
      }
    }
  }
}
```

**Comment by danelowe** [comment]
Result 

```
{
  "took" : 12,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 4,
      "relation" : "eq"
    },
    "max_score" : null,
    "hits" : [ ]
  },
  "aggregations" : {
    "my_sig_text_excluded" : {
      "doc_count" : 4,
      "bg_count" : 8,
      "buckets" : [
        {
          "key" : "storage",
          "doc_count" : 4,
          "score" : 1.0,
          "bg_count" : 4
        }
      ]
    },
    "my_sig_text" : {
      "doc_count" : 4,
      "bg_count" : 8,
      "buckets" : [
        {
          "key" : "storage",
          "doc_count" : 4,
          "score" : 1.0,
          "bg_count" : 4
        },
        {
          "key" : "cloud",
          "doc_count" : 3,
          "score" : 0.375,
          "bg_count" : 4
        }
      ]
    },
    "my_sig_text_included" : {
      "doc_count" : 4,
      "bg_count" : 8,
      "buckets" : [
        {
          "key" : "cloud",
          "doc_count" : 3,
          "score" : 0.375,
          "bg_count" : 4
        }
      ]
    }
  }
}
```

**Comment by imotov** [comment]
Sorry about that. I tested it with 2 development branches and couldn't reproduce the issue. The issue was indeed there in 7.9 but it was fixed in #64520. So it works as expected in master and 7.10.0. Closing.
